src/audio/codec_adapter/codec/cadence.c

```diff
@@ -445,6 +445,13 @@ int cadence_codec_prepare(struct comp_dev *dev)
 		goto free;
 	}
 
+	ret = cadence_codec_init_process(dev);
+	if (ret != LIB_NO_ERROR) {
+		comp_err(dev, "cadence_codec_prepare() error %x: failed to init process",
+			 ret);
+		goto free;
+	}
+
 	comp_dbg(dev, "cadence_codec_prepare() done");
```

Comment:
I don't think this will work. 

``cadence_codec_init_prcess`` will call ``API_CALL(cd, XA_API_CMD_INIT, XA_CMD_TYPE_INIT_PROCESS, NULL, ret);`` 

This will fail because at cadence_codec_prepare() time there is no data in the input buffer. Does this work for Intel DAP codec?

Reply:
@dbaluta but do you think `XA_CMD_TYPE_INIT_PROCESS` requires data in the input buffer?

Reply:
@mrajwa yes. Here is what DAP VLLDP Programmers Guide says:

![image](https://user-images.githubusercontent.com/62316/120198714-18c6dc00-c22b-11eb-9cf4-dd45d5d66c9b.png)

Same for mp3, aac or any other cadence codec. That's why i added the cadence_codec_init_process() call inside the copy() function

Reply:
We need to clarify this with cadence, namely what is the correct way of performing the reset - this is nowhere defined. I noticed INIT_PROCESS does reset the parameters by DAP lib code inspection.

> Does this work for Intel DAP codec?

Yes it works for DAP

Reply:
Ok, i see. It doesn't work for mp3 for example. It needs to have some input data in the buffers. Care to start a thread with Cadence support?
I will be on vacation starting tomorrow returning on June, 9 but will still check my emails.

Reply:
Sure, no problem

Reply:
@dbaluta @mrajwa what's the plan for merging ? do we want to merge the other patches in this PR and wait until clarification from Cadence on this one ?

Reply:
Hi @mrajwa @dbaluta as fixing of the #4104 is still in our v1.8 milestone which should be code freeze by next week, please let us know about any update and if the fix can get aligned and merged before that.

Reply:
@lgirdwood @keyonjie patch updated. I would like to ask @kkarask to verify if it fixes the issue https://github.com/thesofproject/sof/issues/4104

Reply:
@paulstelian97 @dbaluta @iuliana-prodan can someone check with MP3

Reply:
@lgirdwood will verify them today.

Reply:
@mrajwa @lgirdwood the solution is pretty clever. Anyhow, for some strange reasons:

```
+       API_CALL(cd, XA_API_CMD_INIT, XA_CMD_TYPE_INIT_DONE_QUERY,
+                &codec->cpd.init_done, ret);
```

makes codec->cpd.init = 1 and ret = 0!! And this will cause MP3 initialization to be skipped in copy function. 

I think either we have in the API or do we need to do additional steps before acquiring the status. I will send an email to Cadence to ask for this. 

Meanwhile, I think we can also make a per codec function that will initialize codec->cpd.init = 0 or 1.

@lgirdwood @mrajwa will add you in the discussion with Cadence.

Reply:
> Hmm and the playback fails at process() then? If so this is weird and suggest me a codec bug, let's face it, codec returns INIT_DONE but then fails due to uninitialized codec?

Indeed this happens!

Reply:
Can you, for the testing purposes call the codec API with `XA_CMD_TYPE_DO_RUNTIME_INIT` command before the update of init_done at the end of the prepare() function? I have a feeling this may update the flag for compress streams.

Reply:
@mrajwa Ack. Will get back to you when get some time for testing.

Reply:
@mrajwa I checked with your suggestion. But no luck. The init_done is still set at 1.

```
      API_CALL(cd, XA_API_CMD_EXECUTE, XA_CMD_TYPE_DO_RUNTIME_INIT, NULL, ret);
     API_CALL(cd, XA_API_CMD_INIT, XA_CMD_TYPE_INIT_DONE_QUERY, &codec->cpd.init_done, ret);
```
I checked with Cadence support team and they confirm that this happens and might be a library bug. They will forward the problem to engineering. 

Anyhow, what is exactly our problem to fix with this reset procedure. I'm sorry I'm still after vacation and is hard for me to navigate through the sea of emails. Can you provide a simple description of the problem we are trying to solve. 

On another note, I'm having a problem with second cplay run. It's always get stuck. Anyhow, if I leave enough time between 1st run and 2nd run of cplay everything works (system runtime suspend/resume get into action).

Cc: @paulstelian97

Reply:
> On another note, I'm having a problem with second cplay run. It's always get stuck. Anyhow, if I leave enough time between 1st run and 2nd run of cplay everything works (system runtime suspend/resume get into action).

Sounds like kernel runtime PM is rebooting FW and reseting the state ?

Reply:
@lgirdwood true.